README
Tk::GraphMan version 0.02
========================

Tk::GraphMan creates a scrolling Windows Task Manager style grid/graph chart (new values are added to the right-side of the graph).
This is a Tk::Canvas widget that has the look and feel of the Windows Task Manager grid/graph chart.

Example scripts are located in the examples directory.
- Tk-GraphMan-Win32_SysInfo.pl uses the Win32::SystemInfo::CpuUsage and Win32::SystemInfo modules to graph CPU and memory usage in a chrome-less window (no title bar)
                               There are some added bells and whistles which can be accessed by right-clicking on the graph.
                               The graph is also draggable.

- Tk-GraphMan-basic_example.pl uses the rand() function to graph random numbers

Comments and suggestions are always welcome ( bcarroll@cpan.org )

INSTALLATION

To install this module type the following:

     gunzip -c <dist_file>.tar.gz | tar xvf -
     cd <dist_directory>
     perl Makefile.PL
     make test
     make install

Optionally, manually install the module by copying the lib/GraphMan.pm file to /perl/site/lib/Tk/GraphMan.pm (C:\perl\site\lib\Tk\GraphMan.pm on Windows)


DEPENDENCIES

This module requires these other modules and libraries:

  Tk

COPYRIGHT AND LICENCE

Copyright (C) 2010-2014 Brett Carroll

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.6 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Tk::GraphMan.

0.01  Wed Jan  6 11:22:48 2010
  - original version;

0.02  Wed Jul 2 11:37:23 2014
  - fixed directory structure to resolve cpan.pl installation errors
  - fixed Makefile.PL to resolve cpan.pl installation errors
  - added additional example

t/test.pl
# Before `make install' is performed this script should be runnable with
# `make test'. After `make install' it should work as `perl t/test.pl'

######################### We start with some black magic to print on failure.


BEGIN { $| = 1; print "1..1\n"; }
END {print "not ok 1\n" unless $loaded;}
use Tk::GraphMan;
$loaded = 1;
print "ok 1\n";

######################### End of black magic.

#Test 2:
use Tk;
my $main = new MainWindow();
my $grid1 = $main->GraphMan( -interval => 250 )->pack();
my $result = ref($grid1);
print ($result eq "Tk::Canvas" ? "ok 2\n" : "not ok 2\n");




demo/GraphMan.pl
use warnings;
use strict;
use Tk;
use Tk::Menu;
use Win32::SystemInfo::CpuUsage;
use Win32::SystemInfo;

my $X = "233";
my $Y = "103";
my $winX = 10;
my $winY = 20;
my $dragFromX = 0;
my $dragFromY = 0;
my $isDragging = 0;

my $stayontop = 0;
my $showvalues = 0;
my $linecolor = 'green';

my $main = MainWindow->new();
$main->overrideredirect(1);
my $timeline = $main->Canvas( -background => 'black', -height => 100,-width => 230, -relief => 'sunken' )->pack();

my $menu = $main->Menu(-tearoff => 0);
my $CPU = 1;
my $RAM = 1;
my $monitor = $menu->cascade(-label => 'Monitor', -tearoff => 0, -menuitems => [
	[Checkbutton => 'CPU', -variable => \$CPU],
	[Checkbutton => 'RAM', -variable => \$RAM],
	]);
$menu->Checkbutton( -label => 'Show ~Values', -variable => \$showvalues),
$menu->Checkbutton( -label => 'Stay on ~Top', -variable => \$stayontop, -command => sub{ stayontop(); }),
$menu->Separator();
$menu->Button(-label => '~Exit', -command => sub{ $main->destroy(); });

my $position = 11;
my (@cpudata, @ramdata);

&initialize();

my $prevObject = "";
my ($cpu, %mHash, $ram);
my $delay = 1000;
my $timer = $main->repeat($delay, sub {
	if ($CPU){
		$cpu = Win32::SystemInfo::CpuUsage::getCpuUsage(100);
		shift @cpudata;
		push(@cpudata, $cpu);
	}
	if ($RAM){
		Win32::SystemInfo::MemoryStatus(%mHash);
		$ram = $mHash{MemLoad};
		shift @ramdata;
		push(@ramdata, $ram);
	}
	&graph($timeline);
});

$main->bind('<3>', sub { my ($x, $y) = $main->pointerxy; &showmenu($x,$y) } );
$main->bind ('<ButtonPress-1>', \&buttonPress);
$main->bind ('<ButtonRelease-1>', \&buttonRelease);
$main->bind ('<Motion>', \&motion);

$main->update;

MainLoop;

sub graph{
	my $timeline = shift;
	$timeline->delete('grid');
	$timeline->delete('cpudata');
	$timeline->delete('ramdata');
	$timeline->delete('cpuvalues');
	$timeline->delete('ramvalues');
	
	$position = 11 if $position < 1;
	
	my $i;
	for($i=11;$i<230;$i+=11){
		$timeline->createLine( 0, $i, 230, $i, -tags => 'grid', -fill => '#008040', -width => '1' );
	}

	for($i=$position;$i<230;$i+=11){
		$timeline->createLine( $i, 0, $i, 100, -tags => 'grid', -fill => '#008040', -width => '1' );
	}
	
	my ($pos, $lastpoint);
	if ($CPU){
		#graph cpu data
		$pos = 1;
		$lastpoint = 0;
		foreach my $datapoint (@cpudata){
			$timeline->createLine( $pos, 100-$lastpoint, $pos+2, 100-$datapoint, -tags => 'cpudata', -fill => 'green', -width => '1');
			if ($showvalues){
				$timeline->delete('cpuvalues');
				$timeline->createLine( 10, 6, 10, 15, -tags => 'cpuvalues', -fill => 'green', -width => '3');
				$timeline->createText( 25, 10, -text => 'CPU', -tags => 'cpuvalues', -fill => 'white');
				$timeline->createText( 50, 10, -text => $datapoint . '%', -tags => 'cpuvalues', -fill => 'white');
			}
			$pos+=2;
			$lastpoint = $datapoint;
		}
	}
	
		if ($RAM){
		#graph ram data
		$pos = 1;
		$lastpoint = 0;
		foreach my $datapoint (@ramdata){
			$timeline->createLine( $pos, 100-$lastpoint, $pos+2, 100-$datapoint, -tags => 'ramdata', -fill => 'yellow', -width => '1');
			if ($showvalues){
				$timeline->delete('ramvalues');
				$timeline->createLine( 10, 16, 10, 25, -tags => 'ramvalues', -fill => 'yellow', -width => '3');
				$timeline->createText( 25, 20, -text => 'RAM', -tags => 'ramvalues', -fill => 'white');
				$timeline->createText( 50, 20, -text => $datapoint . '%', -tags => 'ramvalues', -fill => 'white');
			}
			$pos+=2;
			$lastpoint = $datapoint;
		}
	}
	$position -= 2;
}

sub showmenu {
  my ($x, $y) = @_;
  $menu->post($x, $y);  # Show the popup menu
}

sub buttonPress {
	$isDragging = 1;
	$dragFromX = $Tk::event->X - $winX;
	$dragFromY = $Tk::event->Y - $winY;
}

sub buttonRelease {
	$isDragging = 0;
}

sub motion {
	return unless $isDragging;
	my $curX = $Tk::event->X;
	my $curY = $Tk::event->Y;
	$curX -= $dragFromX;
	$curY -= $dragFromY;
	$winX = $curX;
	$winY = $curY;
	$main->geometry($X . "x$Y+$curX+$curY");
}

sub initialize{
	#initialize @cpudata and @ramdata arrays with 115 zeros (half the width of the graph)
	for (my $i=1;$i<=115;$i++){
		push (@cpudata, '0');
	}
	
	for (my $i=1;$i<=115;$i++){
		push (@ramdata, '0');
	}
	
	&graph($timeline);
}

sub stayontop{
	$main->attributes(-topmost => $stayontop);
}

examples/Tk-GraphMan-Win32_SysInfo.pl
use warnings;
use strict;
use Tk;
use Tk::Menu;
use Win32::SystemInfo::CpuUsage;
use Win32::SystemInfo;

my $X = "233";
my $Y = "103";
my $winX = 10;
my $winY = 20;
my $dragFromX = 0;
my $dragFromY = 0;
my $isDragging = 0;

my $stayontop = 0;
my $showvalues = 0;
my $linecolor = 'green';

my $main = MainWindow->new();
$main->overrideredirect(1);
my $timeline = $main->Canvas( -background => 'black', -height => 100,-width => 230, -relief => 'sunken' )->pack();

my $menu = $main->Menu(-tearoff => 0);
my $CPU = 1;
my $RAM = 1;
my $monitor = $menu->cascade(-label => 'Monitor', -tearoff => 0, -menuitems => [
  [Checkbutton => 'CPU', -variable => \$CPU],
  [Checkbutton => 'RAM', -variable => \$RAM],
  ]);
$menu->Checkbutton( -label => 'Show ~Values', -variable => \$showvalues),
$menu->Checkbutton( -label => 'Stay on ~Top', -variable => \$stayontop, -command => sub{ stayontop(); }),
$menu->Separator();
$menu->Button(-label => '~Exit', -command => sub{ $main->destroy(); });

my $position = 11;
my (@cpudata, @ramdata);

&initialize();

my $prevObject = "";
my ($cpu, %mHash, $ram);
my $delay = 1000;
my $timer = $main->repeat($delay, sub {
  if ($CPU){
    $cpu = Win32::SystemInfo::CpuUsage::getCpuUsage(100);
    shift @cpudata;
    push(@cpudata, $cpu);
  }
  if ($RAM){
    Win32::SystemInfo::MemoryStatus(%mHash);
    $ram = $mHash{MemLoad};
    shift @ramdata;
    push(@ramdata, $ram);
  }
  &graph($timeline);
});

$main->bind('<3>', sub { my ($x, $y) = $main->pointerxy; &showmenu($x,$y) } );
$main->bind ('<ButtonPress-1>', \&buttonPress);
$main->bind ('<ButtonRelease-1>', \&buttonRelease);
$main->bind ('<Motion>', \&motion);

$main->update;

MainLoop;

sub graph{
  my $timeline = shift;
  $timeline->delete('grid');
  $timeline->delete('cpudata');
  $timeline->delete('ramdata');
  $timeline->delete('cpuvalues');
  $timeline->delete('ramvalues');

  $position = 11 if $position < 1;

  my $i;
  for($i=11;$i<230;$i+=11){
    $timeline->createLine( 0, $i, 230, $i, -tags => 'grid', -fill => '#008040', -width => '1' );
  }

  for($i=$position;$i<230;$i+=11){
    $timeline->createLine( $i, 0, $i, 100, -tags => 'grid', -fill => '#008040', -width => '1' );
  }

  my ($pos, $lastpoint);
  if ($CPU){
    #graph cpu data
    $pos = 1;
    $lastpoint = 0;
    foreach my $datapoint (@cpudata){
      $timeline->createLine( $pos, 100-$lastpoint, $pos+2, 100-$datapoint, -tags => 'cpudata', -fill => 'green', -width => '1');
      if ($showvalues){
        $timeline->delete('cpuvalues');
        $timeline->createLine( 10, 6, 10, 15, -tags => 'cpuvalues', -fill => 'green', -width => '3');
        $timeline->createText( 25, 10, -text => 'CPU', -tags => 'cpuvalues', -fill => 'white');
        $timeline->createText( 50, 10, -text => $datapoint . '%', -tags => 'cpuvalues', -fill => 'white');
      }
      $pos+=2;
      $lastpoint = $datapoint;
    }
  }

    if ($RAM){
    #graph ram data
    $pos = 1;
    $lastpoint = 0;
    foreach my $datapoint (@ramdata){
      $timeline->createLine( $pos, 100-$lastpoint, $pos+2, 100-$datapoint, -tags => 'ramdata', -fill => 'yellow', -width => '1');
      if ($showvalues){
        $timeline->delete('ramvalues');
        $timeline->createLine( 10, 16, 10, 25, -tags => 'ramvalues', -fill => 'yellow', -width => '3');
        $timeline->createText( 25, 20, -text => 'RAM', -tags => 'ramvalues', -fill => 'white');
        $timeline->createText( 50, 20, -text => $datapoint . '%', -tags => 'ramvalues', -fill => 'white');
      }
      $pos+=2;
      $lastpoint = $datapoint;
    }
  }
  $position -= 2;
}

sub showmenu {
  my ($x, $y) = @_;
  $menu->post($x, $y);  # Show the popup menu
}

sub buttonPress {
  $isDragging = 1;
  $dragFromX = $Tk::event->X - $winX;
  $dragFromY = $Tk::event->Y - $winY;
}

sub buttonRelease {
  $isDragging = 0;
}

sub motion {
  return unless $isDragging;
  my $curX = $Tk::event->X;
  my $curY = $Tk::event->Y;
  $curX -= $dragFromX;
  $curY -= $dragFromY;
  $winX = $curX;
  $winY = $curY;
  $main->geometry($X . "x$Y+$curX+$curY");
}

sub initialize{
  #initialize @cpudata and @ramdata arrays with 115 zeros (half the width of the graph)
  for (my $i=1;$i<=115;$i++){
    push (@cpudata, '0');
  }

  for (my $i=1;$i<=115;$i++){
    push (@ramdata, '0');
  }

  &graph($timeline);
}

sub stayontop{
  $main->attributes(-topmost => $stayontop);
}

lib/Tk/GraphMan.pm
package Tk::GraphMan;

use strict;
use vars qw($VERSION);
$VERSION = '0.02';

require Tk;
use Tk::widgets qw(Canvas);
use base qw(Tk::Derived Tk::Canvas);

Construct Tk::Widget 'GraphMan';

sub ClassInit {
  my($class, $mw) = @_;
  $class->SUPER::ClassInit($mw);
}

sub new{
  my ($class, $mw, %args)     = @_;
  bless (\%args, $class);
  my $self                    = $mw->Canvas();
  $self->{'bgcolor'}          = $args{-bgcolor}          || 'black';
  $self->{'grid_color'}       = $args{-grid_color}       || '#008040';
  $self->{'height'}           = $args{-height}           || 100;
  $self->{'width'}            = $args{-width}            || 230;
  #$self->{'relief'}          = $args{-relief}            || 'sunken';
  $self->{'data_title'}       = $args{-data_title}       || 'Data';
  $self->{'data_color'}       = $args{-data_color}       || 'green';
  $self->{'text_color'}       = $args{-text_color}       || 'white';
  $self->{'showvalues'}       = $args{-show_values}      || 'off';
  $self->{'centerline_color'} = $args{-centerline_color} || '#ff0000';
  $self->{'centerline'}       = $args{-centerline}       ||'off';
  $self->{'interval'}         = $args{-interval}         || 1000;
  $self->{'position'}         = 0;

  #initialize @grid_data array with half the width of the graph number of zeros
  for (my $i=1;$i<=($self->{'width'}/2);$i++){
    push (@{ $self->{'grid_data'} }, '0');
  }

  $self->configure(-width => $self->{'width'}, -height => $self->{'height'}, -relief => 'sunken', -background => $self->{'bgcolor'});
  &graph($self);
  return $self;
}

sub UNIVERSAL::start{
  #start scrolling the chart
  my ($self, %args) = @_;
  &graph($self); #start graphing immediately, otherwise graphing will start after $self->{'interval'}
  $self->{'timer'} = $self->repeat($self->{'interval'}, sub {
    &graph($self);
  });
}

sub UNIVERSAL::stop{
  #stop scrolling the chart
  my ($self, %args) = @_;
  $self->{'timer'}->cancel;
}

sub UNIVERSAL::data{
  my ($self, $data) = @_;
  shift @{ $self->{'grid_data'} };
  push (@{ $self->{'grid_data'} }, $data);
}

sub UNIVERSAL::data_array{
  my ($self, @data) = @_;
  shift @{ $self->{'grid_data'} }, scalar(@data);
  push (@{ $self->{'grid_data'} }, @data);
}

sub graph{
  my ($self, %args) = @_;
  $self->delete('grid');
  $self->delete('grid_data');
  $self->delete('grid_values');

  if ($self->{'centerline'} eq 'on'){
    $self->createLine( 0, $self->{'height'}/2, $self->{'width'}, $self->{'height'}/2, -tags => 'center_line', -fill => $self->{'centerline_color'}, -width => '1' );
  } else {
    $self->delete('center_line');
  }

  $self->{'position'} = 11 if $self->{'position'} < 1;

  my $i;
  for($i=11;$i<$self->{'width'};$i+=11){
    $self->createLine( 0, $i, $self->{'width'}, $i, -tags => 'grid', -fill => $self->{'grid_color'}, -width => '1' );
  }

  for($i=$self->{'position'};$i<$self->{'width'};$i+=11){
    $self->createLine( $i, 0, $i, $self->{'height'}, -tags => 'grid', -fill => $self->{'grid_color'}, -width => '1' );
  }

  if ($self->{'centerline'} eq 'on'){
    $self->createLine( 0, $self->{'height'}/2, $self->{'width'}, $self->{'height'}/2, -tags => 'center_line', -fill => $self->{'centerline_color'}, -width => '1' );
  } else {
    $self->delete('center_line');
  }

  my ($pos, $lastpoint);
  #graph grid data
  $pos = 1;
  $lastpoint = 0;
  foreach my $datapoint (@{ $self->{'grid_data'} }){
    $self->createLine( $pos, 100-$lastpoint, $pos+2, 100-$datapoint, -tags => 'grid_data', -fill => $self->{'data_color'}, -width => '1');
    if ($self->{'showvalues'} eq 'on'){
      $self->delete('grid_values');
      $self->createLine( 10, 6, 10, 15, -tags => 'grid_values', -fill => $self->{'data_color'}, -width => '3');
      $self->createText( 25, 10, -text => $self->{'data_title'}, -tags => 'grid_values', -fill => $self->{'text_color'});
      $self->createText( 50, 10, -text => $datapoint . '%', -tags => 'grid_values', -fill => $self->{'text_color'});
    }
    $pos+=2;
    $lastpoint = $datapoint;
  }
  $self->{'position'} -= 2;
  shift @{ $self->{'grid_data'} };
  &pad_grid_data($self);
}

sub pad_grid_data{
  my ($self, %args) = @_;
  while (scalar( @{ $self->{'grid_data'} } ) < ($self->{'width'}/2) ){
    push (@{ $self->{'grid_data'} }, '0');
  }
}

1;
__END__

=head1 NAME

TK::GraphMan - A scrolling Windows Task Manager style grid/graph chart

=head1 SYNOPSIS

  use Tk;
  use Tk::GraphMan;
  my $main = new MainWindow();
  my $grid1 = $main->GraphMan( -interval => 250 )->pack();
  $grid1->Tk::GraphMan::start();
  #do something to generate numeric data for graphing to the chart...
  $grid1->Tk::GraphMan::data(100); #add an integer to the graph's timeline, in scalar context
  # or ...
  $grid1->Tk::GraphMan::data_array(35,36,37,38,39,40,38,36,33); #add an integer to the graph's timeline, in array context
  $main->MainLoop();

=head1 REQUIRES

  Tk

=head1 DESCRIPTION

A scrolling Windows Task Manager style grid/graph chart.  This is a Tk::Canvas widget that has the look and feel of the Windows Task Manager grid/graph chart.

=head1 WIDGET-SPECIFIC OPTIONS

  -bgcolor
  Specify the background color of the chart.  (Default is 'black')

  -grid_color
  Specify the vertical and horizontal grid line colors of the chart.  (Default is '#008040')

  -data_color
  Color of the data being displayed plotted on the chart.  (Default is 'green')

  -interval(milliseconds)
  This options determines how frequently to scroll the chart.

  -height
  Sets the height of the chart, in pixels.  (Default is 100)

  -width
  Sets the width of the chart, in pixels.  (Default is 230)

  -data_title
  Name of the data being displayed.  This is only used when the -showvalues option is set to 'on'.  (Default is 'Data')

  -text_color
  Sets the text color of the -data_title option.  This is only used when the -showvalues option is set to 'on'.  (Default is 'white')

  -showvalues('on'|'off')
  This option toggles printing of the current value (value on the far right of the chart) on and off.  (Default is 'off')

  -centerline('on'|'off')
  This option toggles a line to be drawn horizontally in the center of the graph.  (Default is 'off')

  -centerline_color
  Sets the color of the centerline.  This is only used when the -centerline option is set to 'on'.  (Default is '#ff0000')

=head1 METHODS

=head2 start()

  Begin scrolling the chart.  The chart scrolls to the left, just like the Windows Task Manager chart.

=head2 stop()

  Stop scrolling the chart.

=head2 data($data)

  Plot a number stored in $data scalar on the chart.  Value will be added to the right-side of the chart.

=head2 data_array(@data)

  Plot a list of numbers stored in @data array on the chart.  Values will be added to the right-side of the chart.

=head1 AUTHOR

Brett Carroll, E<lt>bcarroll@cpan.orgE<gt>

=cut
